### ISSUE TITLE
[FEATURE REQ] Fetch Cookie Data

### ISSUE THREAD

**Issue by sujanchegu** [pesos/browser-history#55]
**Is your feature request related to a problem? Please describe.**
Add implementation to fetch cookie data from the sql files.

**Describe the solution you'd like**
A Method in `generic.Outputs` class that can be called to get a dictionary / list / any other data structure that stores the cookies.

**Describe alternatives you've considered**
None

**Additional context**
A security implication as cookies often store data that is not generally expected to be read by the user. 

Some browsers might encrypt the cookies using basic base64 or the like encryption systems which is easy to decrypt using the `base64` package of Py.

**Comment by lucasace** [comment]
I'd like to take this up,
This is what my proposed output would be
`name,url,expiry_date,is_secure,isHttpsonly`

If its okay with you, I can start

**Comment by Samyak2** [comment]
> `name,url,expiry_date,is_secure,isHttpsonly`

@lucasace that's good, but what about the cookie content itself?

**Comment by lucasace** [comment]
> > `name,url,expiry_date,is_secure,isHttpsonly`
> 
> @lucasace that's good, but what about the cookie content itself?

My bad , `name,url,value,expiry_date,is_secure,isHttpsonly`
where value has the cookie content

**Comment by lucasace** [comment]
Sorry for the delay! but unfortunately the delay would have to go for another week
I'll get the PR in by the next weekend if possible

**Comment by Samyak2** [comment]
No issues. Thank you for updating.

**Comment by lucasace** [comment]
I could use some insights on what the decrypted cookie value should contain 
Firefox examples of value : GA1.2.494193081.1602997161 , AJi4QfHGyrvqshYTUlOqnx-OPHe2eAEZU3kBdQDXVM6jUVqY4a09G7pX7P7iqbGvZCAZFm_X8Hs and 2020-10-24-06

Chrome encrypts its cookies using AES Encryption so this is a bit tricky on decrypting  
I get 
13f79712bf2969f2d2f8685cc6d3cc66 ,  true, 
Some decryptions make sense , some dont
As suggested by the issue descritpion i tried decoding these values with base64 decode it gives me something like
b'\xd7w\xfb\xf7\xbdvm\xfd\xbd\xeb\xd7\xf6wg\xfc\xeb\xce\\s\xa7wq\xce\xba'
which doesnt decode using utf-8 and gives weird characters if used with other formats like latin-1

**Comment by Samyak2** [comment]
> GA1.2.494193081.1602997161

That looks like a Google Advertising ID (GAID).

I looked at some values from my installation of firefox and it seems that the values are stored different by each website. Some websites encrypt it while some don't (like the plain date you showed in firefox examples). I don't think we need to decrypt values from firefox as each website would use a different encryption.

---

I wasn't able to get cookie values from Chromium. `select host_key, name, value from cookies;` gave empty values in the `value` column.

Can you show the query or output you're getting for Chromium so that I can take a look at it?

**Comment by lucasace** [comment]
> > GA1.2.494193081.1602997161
> 
> That looks like a Google Advertising ID (GAID).
> 
> I looked at some values from my installation of firefox and it seems that the values are stored different by each website. Some websites encrypt it while some don't (like the plain date you showed in firefox examples). I don't think we need to decrypt values from firefox as each website would use a different encryption.
> 
> I wasn't able to get cookie values from Chromium. `select host_key, name, value from cookies;` gave empty values in the `value` column.
> 
> Can you show the query or output you're getting for Chromium so that I can take a look at it?

The query is Select encrypted_value instead of select value

**Comment by Samyak2** [comment]
According to the answers [here](https://stackoverflow.com/questions/22532870/encrypted-cookies-in-chrome) and [here](https://stackoverflow.com/questions/60416350/chrome-80-how-to-decode-cookies), decrypting this value will be tricky.

According to [this](https://stackoverflow.com/questions/23153159/decrypting-chromium-cookies/23727331#23727331) answer, decrypting it required the user's password.

**Comment by lucasace** [comment]
> According to the answers [here](https://stackoverflow.com/questions/22532870/encrypted-cookies-in-chrome) and [here](https://stackoverflow.com/questions/60416350/chrome-80-how-to-decode-cookies), decrypting this value will be tricky.
> 
> According to [this](https://stackoverflow.com/questions/23153159/decrypting-chromium-cookies/23727331#23727331) answer, decrypting it required the user's password.

I actually found a repo called [browser_cookie3](https://github.com/borisbabic/browser_cookie3) whose code helped me to decrypt it . Decrypted examples are 13f79712bf2969f2d2f8685cc6d3cc66 , true. My question was if u want me to decrypt it further?

**Comment by Samyak2** [comment]
Ah, those decrypted examples are good enough IMO, further decryption will be website specific.

I have one concern though, browser-history has zero dependencies other than python and I would like to keep it that way. Using `browser_cookie3` would add a dependency that is only required for a small secondary feature. 
The code in `browser_cookie3` also seems to be a little hacky and it repeats the platform, history path, etc. detection but in a less elegant way (the profile is hardcoded to `Default`).

If it isn't possible without that package, we should import that module only when required i.e., when user asks for cookies on chrome/chromium and show a message to install the package if the import fails.

If you're considering any alternatives or if I got this wrong, please let me know.

**Comment by lucasace** [comment]
Aah don't worry, i wont use the library,
I'm just using the decryption code
Some dependencies are OS dependent but i think i can handle them . Only concern is pylint warnings. I'll see about them later though
Also a question what and all should i run before making a PR?
Pylint pytest i know
Anything else?

**Comment by Samyak2** [comment]
> Aah don't worry, i wont use the library,
> I'm just using the decryption code
> Some dependencies are OS dependent but i think i can handle them . Only concern is pylint warnings. I'll see about them later

That's great.



> Also a question what and all should i run before making a PR?
> Pylint pytest i know
> Anything else?

A pre-commit hook was added in #88. From the [contributing guidelines](https://browser-history.readthedocs.io/en/latest/contributing.html):

> Initialize the pre-commit hook, if you are committing changes for the first time:
>
>    `pre-commit install` - to setup git hook scripts that will help you check your code each time you commit.

(You might have to do `pip install pre-commit` if it isn't installed)

The pre-commit hook uses flake8 to check code and black to re-format it. It ensures that we don't waste too much time discussing about the formatting. You can run all the checks using `pre-commit run -a` if you want to run them manually.

**Comment by lucasace** [comment]
Some of my [tests](https://github.com/lucasace/browser-history/actions) are failing due to the decryption imports , I couldnt find a requirements.txt file so a little blank on that,
Also for mypy check do i create a [typeshed](https://github.com/python/typeshed) and add stubs for the imports ?

**Comment by Samyak2** [comment]
I want to keep browser-history a zero-dependencies project. That is the reason there's no requirements.txt file. As I said earlier in this thread:

> If it isn't possible without that package, we should import that module only when required i.e., when user asks for cookies on chrome/chromium and show a message to install the package if the import fails.

Avoid importing an external library at the top. You can add the requirements in a `optional-requirements.txt` file (so that it may not be thought of as actual dependencies).

This should fix the failing tests as you're not using cookies in the tests.

---

As for mypy, is it failing for you? I haven't used mypy before, but I don't think you need to create a typeshed just for this.

**Comment by lucasace** [comment]
> I want to keep browser-history a zero-dependencies project. That is the reason there's no requirements.txt file. As I said earlier in this thread:
> 
> > If it isn't possible without that package, we should import that module only when required i.e., when user asks for cookies on chrome/chromium and show a message to install the package if the import fails.
> 
> Avoid importing an external library at the top. You can add the requirements in a `optional-requirements.txt` file (so that it may not be thought of as actual dependencies).

So in the workflow files do i add a pip install -r optional-requirements.txt file ?
and try except Import Error for those packages in browser.py?

**Comment by lucasace** [comment]
> As for mypy, is it failing for you? I haven't used mypy before, but I don't think you need to create a typeshed just for this.

Yuup, its failing on both my system terminal and tests, 
I did do a PR on of the repos which used mypy recently in that they used typeshed, do i do it?

**Comment by Samyak2** [comment]
> So in the workflow files do i add a pip install -r optional-requirements.txt file ?
> and try except Import Error for those packages in browser.py?

Yes. The modules have to be imported when you need them i.e., when that function (`Chrome.cookies_parser`) is called and raise an appropriate error indicating which packages have to be installed.

**Comment by Samyak2** [comment]
> Yuup, its failing on both my system terminal and tests,
> I did do a PR on of the repos which used mypy recently in that they used typeshed, do i do it?
> 
> ```
> browser_history/browsers.py:11: error: Skipping analyzing 'pbkdf2': found module but no type hints or library stubs
> ```

If I understand it correctly, the `pbkdf2` package does not have type hints which mypy needs, right?.

I couldn't find that module in [typeshed](https://github.com/python/typeshed), so are you suggesting manually adding types to it?

**Comment by lucasace** [comment]
> > Yuup, its failing on both my system terminal and tests,
> > I did do a PR on of the repos which used mypy recently in that they used typeshed, do i do it?
> > ```
> > browser_history/browsers.py:11: error: Skipping analyzing 'pbkdf2': found module but no type hints or library stubs
> > ```
> 
> If I understand it correctly, the `pbkdf2` package does not have type hints which mypy needs, right?.
> 
> I couldn't find that module in [typeshed](https://github.com/python/typeshed), so are you suggesting manually adding types to it?

Yuup, i can add it using stubgen

**Comment by Samyak2** [comment]
> Yuup, i can add it using stubgen

That's great. The generated stub file will be included in this repo?

**Comment by lucasace** [comment]
> > Yuup, i can add it using stubgen
> 
> That's great. The generated stub file will be included in this repo?

It has to be added manually once the type hints are generated but yeah, a example would be [cwltool](https://github.com/common-workflow-language/cwltool)
